File: Solutions/Python/populating_next_right_pointers_in_each_node_ii.py
"""
    [MEDIUM]
    117. Populating Next Right Pointers in Each Node II

    Concepts:
    - binary tree
    - breadth-first search

    Stats:
        Runtime | 35 ms     [Beats 96.16%]
        Memory  | 18.98 MB  [Beats 18.71%]
"""

"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""

class Solution:
    def connect(self, root: 'Node') -> 'Node':
        """
            Level-Order Traversal with extra steps
        """
        if not root:
            return None

        queue = deque([root])
        level = []

        while queue:
            len_q = len(queue)

            if level:
                for i in range(len_q - 1):
                    level[i].next = level[i+1]
                level.clear()

            for _ in range(len_q):
                curr = queue.popleft()

                if curr.left:
                    queue.append(curr.left)
                    level.append(curr.left)
                if curr.right:
                    queue.append(curr.right)
                    level.append(curr.right)

        return root

File: Solutions/Cpp/populating_next_right_pointers_in_each_node_ii.cpp
/*
    [MEDIUM]
    117. Populating Next Right Pointers in Each Node II

    Concepts:
    - binary tree
    - breadth-first search

    Stats:
        Runtime | 11 ms     [Beats 45.65%]
        Memory  | 18.91 MB  [Beats 34.02%]
*/

/*
// Definition for a Node.
class Node {
public:
int val;
Node* left;
Node* right;
Node* next;

Node() : val(0), left(NULL), right(NULL), next(NULL) {}

Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}

Node(int _val, Node* _left, Node* _right, Node* _next)
    : val(_val), left(_left), right(_right), next(_next) {}
};
*/

class Solution {
public:
    Node* connect(Node* root) {
        if (root == nullptr)
            return root;

        std::queue<Node*> q;
        q.push(root);

        int len_q;
        Node* curr;

        while (!q.empty()) {
            len_q = q.size();

            for (int i = 0; i < len_q; i++) {
                curr = q.front();
                q.pop();

                if (i != len_q - 1)
                    curr->next = q.front();

                if (curr->left != nullptr)
                    q.push(curr->left);

                if (curr->right != nullptr)
                    q.push(curr->right);
            }
        }

        return root;
    }
};